backends/bmv2/synthesizeValidField.cpp

```diff
@@ -0,0 +1,143 @@
+/*
+Copyright 2013-present Barefoot Networks, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include "synthesizeValidField.h"
+
+#include <boost/range/combine.hpp>
+#include <boost/range/irange.hpp>
+
+#include "backends/bmv2/helpers.h"
+#include "frontends/common/resolveReferences/referenceMap.h"
+#include "frontends/p4/methodInstance.h"
+#include "frontends/p4/typeChecking/typeChecker.h"
+#include "frontends/p4/typeMap.h"
+#include "ir/ir.h"
+
+namespace BMV2 {
+
+/// Adds a valid bit to all header types.
+class AddValidField final : public Modifier {
+ public:
+    AddValidField() { setName("AddValidField"); }
+ private:
+    bool preorder(IR::Type_Header* header) override {
+        const cstring validField = V1ModelProperties::validField;
+        if (header->getField(validField) != nullptr) return true;
+        auto field = new IR::StructField(validField, IR::Type::Bits::get(1));
+        header->fields.push_back(field);
+        return true;
+    }
+};
+
+/// Replaces calls to isValid() with references to the header's valid bit.
```

Comment:
So I investigated this, and I don't think we're generating correct BMV2 JSON for header unions as it is. The `valid` match type won't work for header unions; we should be making use of the `valid_union` primitive. `valid_union` can't be used in table match keys (since they can't be general expressions) so what we need to do is use `valid_union` to set a scalar and use the scalar in the key. I think the right fix here is:

1. Don't remove `isValid()` calls on header unions in `SynthesizeValidField`. Since all other `isValid()` calls will be removed, these will be the only `isValid()` calls left.
2. Modify the `KeyIsComplex` policies to treat `isValid()` calls as complex, so they're moved out of match keys. (Or perhaps, for now, just add a new policy, and remove the old ones later once we're sure all backends have migrated over.)
3. In the backend, generate the `valid_union` primitive when we encounter `isValid()` calls on header unions.

Should all be pretty straightforward.